File: packages/embedding-engine/src/graph.py
"""
Knowledge graph builder for context-pilot.
Nodes = chunks (functions, classes, modules).
Edges = import relationships, function calls, class inheritance.
"""
from __future__ import annotations

import re
import sqlite3
from pathlib import Path
from typing import Optional

try:
    import networkx as nx
    NX_AVAILABLE = True
except ImportError:
    NX_AVAILABLE = False


# ── Edge extraction from source code ──────────────────────────────────────────

_IMPORT_PATTERNS = [
    # Python: from foo import bar / import foo.bar
    re.compile(r"^(?:from|import)\s+([\w.]+)", re.MULTILINE),
    # TS/JS: import ... from './foo' or require('./foo')
    re.compile(r"""(?:import|require)\s*(?:\(?\s*['"])(\.{1,2}/[\w./]+)['"]"""),
    # TS/JS: export ... from './foo'
    re.compile(r"""export\s+.*?from\s+['\"](\.{1,2}/[\w./]+)['\"]"""),
]

_CALL_PATTERN = re.compile(r"\b(\w+)\s*\(")
_EXTENDS_PATTERN = re.compile(r"(?:extends|implements)\s+(\w+)")


def _extract_imports(content: str) -> list[str]:
    imports = []
    for pattern in _IMPORT_PATTERNS:
        imports.extend(pattern.findall(content))
    return list(set(imports))


def _extract_calls(content: str) -> list[str]:
    calls = _CALL_PATTERN.findall(content)
    # Filter out common keywords and builtins
    skip = {"if", "for", "while", "return", "print", "len", "range", "str",
             "int", "list", "dict", "set", "type", "super", "self", "cls"}
    return [c for c in calls if c not in skip]


def _extract_extends(content: str) -> list[str]:
    return _EXTENDS_PATTERN.findall(content)


# ── Graph construction ─────────────────────────────────────────────────────────

def build_graph(conn: sqlite3.Connection, project_id: str) -> "nx.DiGraph":
    if not NX_AVAILABLE:
        raise RuntimeError("networkx is not installed. Run: pip install networkx")

    G = nx.DiGraph()

    # Load all chunks with their file paths
    rows = conn.execute("""
        SELECT c.id, c.chunk_type, c.name, c.content, f.path
        FROM chunks c
        JOIN files f ON f.id = c.file_id
        WHERE f.project_id = ?
    """, (project_id,)).fetchall()

    # name -> chunk_id index for resolving references
    name_index: dict[str, str] = {}

    for row in rows:
        G.add_node(row["id"], chunk_type=row["chunk_type"], name=row["name"], path=row["path"])
        if row["name"]:
            name_index[row["name"]] = row["id"]

    # Add edges based on imports, calls, and inheritance
    for row in rows:
        content = row["content"]
        chunk_id = row["id"]

        for call in _extract_calls(content):
            target_id = name_index.get(call)
            if target_id and target_id != chunk_id:
                G.add_edge(chunk_id, target_id, edge_type="calls", weight=0.8)

        for base in _extract_extends(content):
            target_id = name_index.get(base)
            if target_id and target_id != chunk_id:
                G.add_edge(chunk_id, target_id, edge_type="extends", weight=1.0)

    return G


def get_subgraph(
    G: "nx.DiGraph",
    target_chunk_id: str,
    depth: int = 2,
    direction: str = "both",
) -> dict:
    if target_chunk_id not in G:
        return {"nodes": [], "edges": [], "error": f"Node not found: {target_chunk_id}"}

    if direction == "outgoing":
        reachable = nx.descendants(G, target_chunk_id)
        subG = G.subgraph({target_chunk_id} | _limit_depth(G, target_chunk_id, depth, "out"))
    elif direction == "incoming":
        reachable = nx.ancestors(G, target_chunk_id)
        subG = G.subgraph({target_chunk_id} | _limit_depth(G, target_chunk_id, depth, "in"))
    else:
        out_nodes = _limit_depth(G, target_chunk_id, depth, "out")
        in_nodes = _limit_depth(G, target_chunk_id, depth, "in")
        subG = G.subgraph({target_chunk_id} | out_nodes | in_nodes)

    nodes = [
        {"id": n, **G.nodes[n]}
        for n in subG.nodes
    ]
    edges = [
        {"from": u, "to": v, **d}
        for u, v, d in subG.edges(data=True)
    ]
    return {"nodes": nodes, "edges": edges}


def _limit_depth(G: "nx.DiGraph", start: str, depth: int, direction: str) -> set[str]:
    visited: set[str] = set()
    frontier = {start}
    for _ in range(depth):
        next_frontier: set[str] = set()
        for node in frontier:
            if direction == "out":
                neighbors = set(G.successors(node))
            else:
                neighbors = set(G.predecessors(node))
            next_frontier |= neighbors - visited - {start}
        visited |= next_frontier
        frontier = next_frontier
    return visited


def get_graph_distances(
    G: "nx.DiGraph", source_chunk_id: str, target_ids: list[str]
) -> dict[str, float]:
    """
    Returns normalized distance [0,1] from source to each target.
    Closer = lower distance = higher relevance for ranking.
    """
    distances: dict[str, float] = {}
    if source_chunk_id not in G:
        return {t: 1.0 for t in target_ids}

    for target in target_ids:
        if target not in G:
            distances[target] = 1.0
            continue
        try:
            # Try both directions — undirected shortest path
            undirected = G.to_undirected()
            d = nx.shortest_path_length(undirected, source_chunk_id, target)
            # Normalize: distance 0=same, 1=very far (cap at 5)
            distances[target] = min(d / 5.0, 1.0)
        except nx.NetworkXNoPath:
            distances[target] = 1.0

    return distances


def find_chunk_by_name_or_path(
    conn: sqlite3.Connection, project_id: str, target: str
) -> Optional[str]:
    """Resolve a file path or function name to a chunk_id."""
    # Try exact function/class name match
    row = conn.execute("""
        SELECT c.id FROM chunks c
        JOIN files f ON f.id = c.file_id
        WHERE f.project_id = ? AND c.name = ?
        LIMIT 1
    """, (project_id, target)).fetchone()
    if row:
        return row["id"]

    # Try file path match (return first chunk of that file)
    row = conn.execute("""
        SELECT c.id FROM chunks c
        JOIN files f ON f.id = c.file_id
        WHERE f.project_id = ? AND f.path LIKE ?
        ORDER BY c.start_line ASC
        LIMIT 1
    """, (project_id, f"%{target}%")).fetchone()
    return row["id"] if row else None

File: packages/engine/src/graph/graph.ts
export interface NodeAttrs {
  chunkType: string;
  name: string | null;
  path: string;
}

export interface EdgeAttrs {
  edgeType: string;
  weight: number;
}

export class DiGraph {
  private nodes = new Map<string, NodeAttrs>();
  private outEdges = new Map<string, Map<string, EdgeAttrs>>();
  private inEdges = new Map<string, Set<string>>();

  addNode(id: string, attrs: NodeAttrs): void {
    this.nodes.set(id, attrs);
    if (!this.outEdges.has(id)) this.outEdges.set(id, new Map());
    if (!this.inEdges.has(id)) this.inEdges.set(id, new Set());
  }

  addEdge(from: string, to: string, attrs: EdgeAttrs): void {
    if (!this.outEdges.has(from)) this.outEdges.set(from, new Map());
    if (!this.inEdges.has(to)) this.inEdges.set(to, new Set());
    this.outEdges.get(from)!.set(to, attrs);
    this.inEdges.get(to)!.add(from);
  }

  has(id: string): boolean {
    return this.nodes.has(id);
  }

  getNode(id: string): NodeAttrs | undefined {
    return this.nodes.get(id);
  }

  successors(id: string): string[] {
    return [...(this.outEdges.get(id)?.keys() ?? [])];
  }

  predecessors(id: string): string[] {
    return [...(this.inEdges.get(id) ?? [])];
  }

  getEdge(from: string, to: string): EdgeAttrs | undefined {
    return this.outEdges.get(from)?.get(to);
  }

  allNodes(): Array<[string, NodeAttrs]> {
    return [...this.nodes.entries()];
  }

  allEdges(): Array<{ from: string; to: string; attrs: EdgeAttrs }> {
    const result: Array<{ from: string; to: string; attrs: EdgeAttrs }> = [];
    for (const [from, tos] of this.outEdges) {
      for (const [to, attrs] of tos) {
        result.push({ from, to, attrs });
      }
    }
    return result;
  }

  nodeCount(): number {
    return this.nodes.size;
  }

  /**
   * BFS subgraph starting from `rootId`, up to `depth` hops.
   * direction: "outgoing" | "incoming" | "both"
   */
  subgraph(
    rootId: string,
    depth: number,
    direction: "outgoing" | "incoming" | "both"
  ): { nodeIds: Set<string>; edges: Array<{ from: string; to: string; attrs: EdgeAttrs }> } {
    const visited = new Set<string>([rootId]);
    const queue: Array<{ id: string; d: number }> = [{ id: rootId, d: 0 }];
    const edges: Array<{ from: string; to: string; attrs: EdgeAttrs }> = [];

    while (queue.length > 0) {
      const { id, d } = queue.shift()!;
      if (d >= depth) continue;

      const neighbors: string[] = [];
      if (direction === "outgoing" || direction === "both") neighbors.push(...this.successors(id));
      if (direction === "incoming" || direction === "both") neighbors.push(...this.predecessors(id));

      for (const neighbor of neighbors) {
        if (!visited.has(neighbor)) {
          visited.add(neighbor);
          queue.push({ id: neighbor, d: d + 1 });
        }
        // collect edges in original direction
        if (this.outEdges.get(id)?.has(neighbor)) {
          edges.push({ from: id, to: neighbor, attrs: this.getEdge(id, neighbor)! });
        } else if (this.outEdges.get(neighbor)?.has(id)) {
          edges.push({ from: neighbor, to: id, attrs: this.getEdge(neighbor, id)! });
        }
      }
    }

    return { nodeIds: visited, edges };
  }

  /**
   * Compute normalized shortest-path distances from `sourceId` to each candidate.
   * Uses BFS on undirected version. Returns 0–1 where 0 = same node, 1 = unreachable.
   */
  shortestPathDistances(sourceId: string, targetIds: string[]): Record<string, number> {
    const dist = new Map<string, number>();
    const queue: Array<{ id: string; d: number }> = [{ id: sourceId, d: 0 }];
    dist.set(sourceId, 0);

    while (queue.length > 0) {
      const { id, d } = queue.shift()!;
      const neighbors = [...this.successors(id), ...this.predecessors(id)];
      for (const n of neighbors) {
        if (!dist.has(n)) {
          dist.set(n, d + 1);
          queue.push({ id: n, d: d + 1 });
        }
      }
    }

    const maxDist = Math.max(1, ...dist.values());
    const result: Record<string, number> = {};

    for (const targetId of targetIds) {
      const d = dist.get(targetId);
      result[targetId] = d === undefined ? 1.0 : Math.min(1.0, d / maxDist);
    }

    return result;
  }
}
